File: ideal-job-scheduler/ideal-job-scheduler-core/src/main/java/cn/sh/ideal/job/scheduler/core/dispatch/handler/impl/BaseHttpExecuteHandler.java
package cn.sh.ideal.job.scheduler.core.dispatch.handler.impl;

import cn.sh.ideal.job.common.constants.HandleStatusEnum;
import cn.sh.ideal.job.common.constants.RouteStrategyEnum;
import cn.sh.ideal.job.common.constants.TriggerTypeEnum;
import cn.sh.ideal.job.common.exception.VisibleException;
import cn.sh.ideal.job.common.http.HttpMethod;
import cn.sh.ideal.job.common.http.HttpRequest;
import cn.sh.ideal.job.common.http.HttpScriptUtils;
import cn.sh.ideal.job.common.utils.JsonUtils;
import cn.sh.ideal.job.common.utils.ReactorUtils;
import cn.sh.ideal.job.scheduler.core.admin.entity.JobInstance;
import cn.sh.ideal.job.scheduler.core.admin.entity.vo.DispatchJobView;
import cn.sh.ideal.job.scheduler.core.admin.service.JobInstanceService;
import cn.sh.ideal.job.scheduler.core.dispatch.handler.ExecuteHandler;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpHeaders;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.reactive.function.BodyInserters;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.ExecutorService;

/**
 * @author 宋志宗
 * @date 2020/9/3
 */
public abstract class BaseHttpExecuteHandler implements ExecuteHandler {
    private static final Logger log = LoggerFactory.getLogger(BaseHttpExecuteHandler.class);
    private final WebClient webClient = ReactorUtils
            .createWebClient(400, 400, 120_000);
    @Nonnull
    private final JobInstanceService instanceService;
    @Nonnull
    private final ExecutorService jobCallbackThreadPool;

    protected BaseHttpExecuteHandler(@Nonnull JobInstanceService instanceService,
                                     @Nonnull ExecutorService jobCallbackThreadPool) {

        this.instanceService = instanceService;
        this.jobCallbackThreadPool = jobCallbackThreadPool;
    }

    @SuppressWarnings("DuplicatedCode")
    @Override
    public final void execute(@Nonnull JobInstance instance,
                              @Nonnull DispatchJobView jobView,
                              @Nonnull TriggerTypeEnum triggerType,
                              @Nullable String customExecuteParam) {
        String executeParam = customExecuteParam;
        if (executeParam == null) {
            executeParam = jobView.getExecuteParam();
        }
        if (StringUtils.isBlank(executeParam)) {
            log.info("任务: {} Http script为空", jobView.getJobId());
            throw new VisibleException("Http script为空");
        }
        HttpRequest httpRequest;
        try {
            httpRequest = HttpScriptUtils.parse(executeParam);
        } catch (HttpScriptUtils.HttpScriptParseException e) {
            log.info("任务: {} http script解析异常: {}", jobView.getJobId(), e.getMessage());
            throw new VisibleException("http script解析异常: ${e.message}");
        }
        RouteStrategyEnum routeStrategy = jobView.getRouteStrategy();
        String url = httpRequest.getUrl();
        List<String> chooseServer = getAddressList(jobView.getJobId(), url, routeStrategy);
        if (chooseServer.isEmpty()) {
            log.info("任务: {} 选取远程服务为空", jobView.getJobId());
            throw new VisibleException("选取远程服务为空");
        }
        MultiValueMap<String, String> formData = new LinkedMultiValueMap<>();
        Object requestBody = null;
        HttpMethod method = httpRequest.getMethod();
        HttpHeaders headers = httpRequest.getHeaders();
        String body = httpRequest.getBody();
        String queryString = httpRequest.getQueryString();
        if (StringUtils.isNoneBlank(body)) {
            if ((StringUtils.startsWith(body, "{")
                    && StringUtils.endsWith(body, "}"))
                    || (StringUtils.startsWith(body, "[")
                    && StringUtils.endsWith(body, "]"))) {
                requestBody = JsonUtils.parseJson(body, Object.class);
            } else {
                String[] singles = StringUtils.split(body, "&");
                assert singles != null;
                for (String single : singles) {
                    String[] sp = StringUtils.split(single, "=", 2);
                    if (sp.length == 2) {
                        formData.add(sp[0], sp[1]);
                    } else {
                        log.info("任务: {} http script body不合法", jobView.getJobId());
                    }
                }
            }
        }
        long handleTime = System.currentTimeMillis();
        if (chooseServer.size() == 1) {
            String uri = chooseServer.get(0);
            String requestUri = StringUtils.isNoneBlank(queryString)
                    ? uri + "?" + queryString : uri;
            WebClient.RequestHeadersSpec<?> clientSpec
                    = buildWebClientSpec(method, requestUri, requestBody, headers, formData);
            Mono<String> bodyToMono = clientSpec.retrieve().bodyToMono(String.class);
            subscriptResponse(bodyToMono, instance, handleTime);
        } else {
            for (String uri : chooseServer) {
                JobInstance jobInstance = JobInstance.createInitialized();
                jobInstance.setParentId(instance.getInstanceId());
                jobInstance.setJobId(jobView.getJobId());
                jobInstance.setExecutorId(jobView.getExecutorId());
                jobInstance.setTriggerType(triggerType);
                jobInstance.setSchedulerInstance(instance.getSchedulerInstance());
                jobInstance.setExecutorHandler(jobView.getExecutorHandler());
                jobInstance.setExecuteParam(executeParam);
                jobInstance.setExecutorInstance(uri);
                String requestUri = StringUtils.isNoneBlank(queryString)
                        ? uri + "?" + queryString : uri;
                WebClient.RequestHeadersSpec<?> clientSpec
                        = buildWebClientSpec(method, requestUri, requestBody, headers, formData);
                Mono<String> bodyToMono = clientSpec.retrieve().bodyToMono(String.class);
                subscriptResponse(bodyToMono, instance, handleTime);
            }
        }
    }

    private void subscriptResponse(@Nonnull Mono<String> bodyToMono,
                                   @Nonnull JobInstance jobInstance,
                                   long handleTime) {
        bodyToMono.onErrorResume(e -> {
            String errMsg = e.getClass().getName() + ": " + e.getMessage();
            log.info("http调度异常: {}", errMsg);
            jobInstance.setHandleStatus(HandleStatusEnum.ABNORMAL);
            return Mono.just(errMsg);
        }).doOnNext(result -> {
            jobInstance.setResult(result);
            if (jobInstance.getHandleStatus() != HandleStatusEnum.ABNORMAL) {
                jobInstance.setHandleStatus(HandleStatusEnum.COMPLETE);
            }
        }).doFinally(signalType -> {
            jobInstance.setHandleTime(handleTime);
            jobInstance.setFinishedTime(System.currentTimeMillis());
            jobInstance.setSequence(2);
            jobCallbackThreadPool.execute(() -> instanceService.saveInstance(jobInstance));
        }).subscribe();
    }

    private WebClient.RequestHeadersSpec<?> buildWebClientSpec(
            @Nonnull HttpMethod method,
            @Nonnull String requestUri,
            @Nullable Object requestBody,
            @Nullable HttpHeaders headers,
            @Nonnull MultiValueMap<String, String> formData) {
        WebClient.RequestHeadersSpec<?> client;
        switch (method) {
            case GET: {
                client = webClient.get().uri(requestUri);
                break;
            }
            case DELETE: {
                client = webClient.delete().uri(requestUri);
                break;
            }
            case POST: {
                WebClient.RequestBodySpec spec = webClient.post().uri(requestUri);
                if (requestBody != null) {
                    spec.body(BodyInserters.fromValue(requestBody));
                } else if (!formData.isEmpty()) {
                    spec.body(BodyInserters.fromFormData(formData));
                }
                client = spec;
                break;
            }
            case PATCH: {
                WebClient.RequestBodySpec spec = webClient.patch().uri(requestUri);
                if (requestBody != null) {
                    spec.body(BodyInserters.fromValue(requestBody));
                } else if (!formData.isEmpty()) {
                    spec.body(BodyInserters.fromFormData(formData));
                }
                client = spec;
                break;
            }
            case PUT: {
                WebClient.RequestBodySpec spec = webClient.put().uri(requestUri);
                if (requestBody != null) {
                    spec.body(BodyInserters.fromValue(requestBody));
                } else if (!formData.isEmpty()) {
                    spec.body(BodyInserters.fromFormData(formData));
                }
                client = spec;
                break;
            }
            default: {
                // 不应该发生的
                String message = "不合法的Http method: " + method.name();
                log.error(message);
                throw new VisibleException(message);
            }
        }
        if (headers != null) {
            client.headers(httpHeaders -> headers.forEach(httpHeaders::put));
        }
        return client;
    }

    protected List<String> getAddressList(long jobId, @Nonnull String scriptUrl,
                                          @Nonnull RouteStrategyEnum routeStrategy) {
        return Collections.singletonList(scriptUrl);
    }
}

File: ideal-job-scheduler/ideal-job-scheduler-core/src/main/kotlin/cn/sh/ideal/job/scheduler/core/dispatch/handler/impl/BaseHttpExecuteHandler.kt
package cn.sh.ideal.job.scheduler.core.dispatch.handler.impl

import cn.sh.ideal.job.common.constants.HandleStatusEnum
import cn.sh.ideal.job.common.constants.RouteStrategyEnum
import cn.sh.ideal.job.common.constants.TriggerTypeEnum
import cn.sh.ideal.job.common.exception.VisibleException
import cn.sh.ideal.job.common.http.HttpMethod
import cn.sh.ideal.job.common.http.HttpScriptUtils
import cn.sh.ideal.job.common.utils.JsonUtils
import cn.sh.ideal.job.scheduler.core.admin.entity.JobInfo
import cn.sh.ideal.job.scheduler.core.admin.entity.JobInstance
import cn.sh.ideal.job.scheduler.core.admin.service.JobInstanceService
import cn.sh.ideal.job.scheduler.core.dispatch.handler.ExecuteHandler
import cn.sh.ideal.job.scheduler.core.utils.WebClients
import org.apache.commons.lang3.StringUtils
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.http.HttpHeaders
import org.springframework.util.LinkedMultiValueMap
import org.springframework.web.reactive.function.BodyInserters
import org.springframework.web.reactive.function.client.WebClient
import reactor.core.publisher.Mono
import java.util.concurrent.ExecutorService

/**
 * @author 宋志宗
 * @date 2020/8/28
 */
abstract class BaseHttpExecuteHandler(
    private val instanceService: JobInstanceService,
    private val jobCallbackThreadPool: ExecutorService) : ExecuteHandler {
  val log: Logger = LoggerFactory.getLogger(this.javaClass)

  private val webClient = WebClients.createWebClient(
      400, 400, 120_000)

  @Suppress("DuplicatedCode")
  override fun execute(instance: JobInstance,
                       jobInfo: JobInfo,
                       triggerType: TriggerTypeEnum,
                       customExecuteParam: String?) {
    val jobId = jobInfo.jobId
    val executeParam = customExecuteParam ?: jobInfo.executeParam
    if (executeParam.isBlank()) {
      log.info("任务: {} Http script为空", jobId)
      throw VisibleException("Http script为空")
    }
    val httpRequest = try {
      HttpScriptUtils.parse(executeParam)
    } catch (e: Exception) {
      log.info("任务: {} http script解析异常: {}", jobId, e.message)
      throw VisibleException("http script解析异常: ${e.message}")
    }
    val routeStrategy = jobInfo.routeStrategy
    val url = httpRequest.url
    val chooseServer = getAddressList(jobId, routeStrategy, url)
    if (chooseServer.isEmpty()) {
      log.info("任务: {} 选取远程服务为空", jobId)
      throw VisibleException("选取远程服务为空")
    }

    val method = httpRequest.method
    val queryString = httpRequest.queryString
    val headers = httpRequest.headers
    val body = httpRequest.body
    val multiValueMap = LinkedMultiValueMap<String, String>()
    var requestBody: Any? = null
    if (body?.isNotBlank() == true) {
      if ((body.startsWith("{") && body.endsWith("}"))
          || (body.startsWith("[") && body.endsWith("]"))) {
        requestBody = JsonUtils.parseJson(body, Any::class.java)
      } else {
        StringUtils.split(body, "&")
            .forEach { single ->
              StringUtils.split(single, "=", 2)
                  .let { sp ->
                    if (sp.size == 2) {
                      multiValueMap.add(sp[0], sp[1])
                    } else {
                      log.info("任务: {} http script body不合法")
                    }
                  }
            }
      }
    }
    val currentTimeMillis = System.currentTimeMillis()
    if (chooseServer.size == 1) {
      val uri = chooseServer[0]
      instance.executorInstance = uri
      val requestUri = if (queryString?.isNotBlank() == true) "$uri:$queryString" else uri
      val client = buildWebClientSpec(
          method, requestUri, requestBody, multiValueMap, headers)
      val bodyToMono = client.retrieve().bodyToMono(String::class.java)
      subscriptResponse(bodyToMono, instance, currentTimeMillis)
    } else {
      for (uri in chooseServer) {
        val jobInstance = JobInstance.createInitialized()
        jobInstance.parentId = instance.instanceId
        jobInstance.jobId = instance.jobId
        jobInstance.executorId = instance.executorId
        jobInstance.triggerType = triggerType
        jobInstance.schedulerInstance = instance.schedulerInstance
        jobInstance.executorHandler = instance.executorHandler
        jobInstance.executeParam = executeParam
        jobInstance.executorInstance = uri
        val requestUri = if (queryString?.isNotBlank() == true) "$uri:$queryString" else uri
        val client = buildWebClientSpec(
            method, requestUri, requestBody, multiValueMap, headers)
        val bodyToMono = client.retrieve().bodyToMono(String::class.java)
        subscriptResponse(bodyToMono, jobInstance, currentTimeMillis)
      }
    }
  }

  private fun subscriptResponse(bodyToMono: Mono<String>,
                                instance: JobInstance,
                                currentTimeMillis: Long) {
    bodyToMono.onErrorResume { e ->
      val errMsg = "${e.javaClass.name}:${e.message}"
      instance.handleStatus = HandleStatusEnum.ABNORMAL
      Mono.just(errMsg)
    }.map { result ->
      instance.result = result
      if (instance.handleStatus != HandleStatusEnum.ABNORMAL) {
        instance.handleStatus = HandleStatusEnum.COMPLETE
      }
    }.doFinally {
      instance.handleTime = currentTimeMillis
      instance.finishedTime = System.currentTimeMillis()
      instance.sequence = 2
      jobCallbackThreadPool.execute { instanceService.saveInstance(instance) }
    }.subscribe()
  }

  private fun buildWebClientSpec(method: HttpMethod,
                                 requestUri: String,
                                 requestBody: Any?,
                                 multiValueMap: LinkedMultiValueMap<String, String>,
                                 headers: HttpHeaders?): WebClient.RequestHeadersSpec<*> {
    val client = when (method) {
      HttpMethod.GET -> webClient.get().uri(requestUri)
      HttpMethod.DELETE -> webClient.delete().uri(requestUri)
      HttpMethod.POST -> {
        val requestBodySpec = webClient.post().uri(requestUri)
        if (requestBody != null) {
          requestBodySpec.body(BodyInserters.fromValue(requestBody))
        } else if (multiValueMap.isNotEmpty()) {
          BodyInserters.fromFormData(multiValueMap)
        }
        requestBodySpec
      }
      HttpMethod.PATCH -> {
        val requestBodySpec = webClient.patch().uri(requestUri)
        if (requestBody != null) {
          requestBodySpec.body(BodyInserters.fromValue(requestBody))
        } else if (multiValueMap.isNotEmpty()) {
          BodyInserters.fromFormData(multiValueMap)
        }
        requestBodySpec
      }
      HttpMethod.PUT -> {
        val requestBodySpec = webClient.put().uri(requestUri)
        if (requestBody != null) {
          requestBodySpec.body(BodyInserters.fromValue(requestBody))
        } else if (multiValueMap.isNotEmpty()) {
          BodyInserters.fromFormData(multiValueMap)
        }
        requestBodySpec
      }
      else -> {
        // 不应该发生的
        val message = "不合法的Http method: $method"
        log.error(message)
        throw VisibleException(message)
      }
    }
    if (headers != null) {
      client.headers {
        headers.forEach { name, values -> it[name] = values }
      }
    }
    return client
  }

  open fun getAddressList(jobId: Long,
                          routeStrategy: RouteStrategyEnum,
                          scriptUrl: String): List<String> {
    return listOf(scriptUrl)
  }
}